src/ir_pattern.py

```diff
@@ -0,0 +1,423 @@
+        pattern_refs = pattern_node.block.instruction_refs
+
+        # Split the pattern asm into 3 disjoint sets of instructions:
+        # input_refs ("in.fictive"s), body_refs, and tail_ref (the last instruction)
+        n_inputs = len(pattern.replacement_instr.inputs)
+        input_refs, body_refs, tail_ref = (
+            pattern_refs[:n_inputs],
+            pattern_refs[n_inputs:-1],
+            pattern_refs[-1],
+        )
+        assert all(r.instruction.mnemonic == "in.fictive" for r in input_refs)
+        assert all(r.instruction.mnemonic != "in.fictive" for r in body_refs)
+
+        # For now, pattern inputs must be Registers, not StackLocations. It's not always
+        # trivial to create temporary StackLocations in the same way we create temporary
+        # Registers during replacement.
+        assert all(
+            isinstance(inp, Register) for inp in pattern.replacement_instr.inputs
+        )
+
+        # For now, patterns can only have 1 output register (which must be set by the
+        # final instruction in the pattern). This simplifies the replacement step because
+        # we can replace the final instruction and know that all the pattern inputs have
+        # been assigned by there, and the output has not yet been used.
+        assert len(pattern.replacement_instr.outputs) == 1
+        assert pattern.replacement_instr.outputs == tail_ref.instruction.outputs
+
+        # Start the matching by finding all possible matches for the last instruction
+        try_matches = []
+        tail_inputs = pattern.flow_graph.instr_inputs[tail_ref]
+        for cand_ref in refs_by_mnemonic.get(tail_ref.instruction.mnemonic, []):
+            state = TryIrMatch()
+            if state.match_refset(RefSet([tail_ref]), RefSet([cand_ref])):
+                try_matches.append(state)
+
+        # Continue matching by working backwards through the pattern
+        for pat_ref in [tail_ref] + body_refs[::-1]:
+            pat_inputs = pattern.flow_graph.instr_inputs[pat_ref]
+
+            next_try_matches = []
+            for state in try_matches:
+                # By pattern construction, pat_ref should be in the state's ref_map
+                # (It would be missing for "disjoint" or irrelevant instructions in the
+                # pattern, like random nops: these aren't allowed)
+                cand = state.try_map_ref(pat_ref)
+                if not isinstance(cand, InstrRef):
+                    continue
+                if not state.match_instr(pat_ref.instruction, cand.instruction):
+                    continue
+                if not state.match_inputrefs(
+                    pat_inputs,
+                    flow_graph.instr_inputs[cand],
+                ):
+                    continue
+                next_try_matches.append(state)
+            try_matches = next_try_matches
+
+        for n, state in enumerate(try_matches):
+            # Perform any additional pattern-specific validation or computation
```

Comment:
Ah yeah, the idea was to compute literals. 
I could have it pass a `TryIrMatch` instead, but I think that'd really weaken the intent behind the `TryIrMatch`/`IrMatch` split.
I'll leave the "computation" comment out for now, since no patterns make use of it yet anyways.